README.md
# ObjectMover
World of Warcraft Addon designed for use on servers running the Enteleaie Core that helps users place, move, and manipulate game objects.

## Installation
Drag the "ObjectMover" addon to your Epsilon/\_retail_/Interface/Addons/ folder.
Launch the game, and login. In bottom left of Character Select, click "Addons" and then ensure ObjectMover is enabled. Use "/om", "/obj" or click on the button on the MiniMap once ingame to show the GUI. (Can be dragged around to reposition anywhere around the minimap).

https://forums.epsilonwow.net/topic/467-addon-convenientcommands-objectmover/


ObjectMover/Changelog.txt
v7.0.0 (July 29, 2021)
	- NEW: Added Tint & Object Spell Support
	- NEW: Object Group Support (Automatic)
	- NEW: Automatic Dimension Detection for M2's
	- NEW: Auto Update mode for Object Parameters (Will auto fill selected objects ID, Scale, and if it's an M2, the dimensions as well)
	- NEW: Auto Show option - Toggle on to automatically show ObjectMover when you load in
	- NEW: Fade Out option - Toggle on to fade ObjectMover to be less distracting when not in immediate use
	- UPDATED: Scale Object now works with Macro Spawning instead of delayed scale commands (No Delay, No Desync)
	- UPDATED: Scale Object now applies to the selected object when entered
	- UPDATED: Saved Presets can now be deleted directly from the UI (Right-Click in Load Dropdown to delete, no need to use chat commands now)
	- CHANGED: Spawn is now in Object section; Revised Movement button placements
	- CHANGED: Added a "Classic Layout" button in Movement to revert the movement buttons to closer to the original layout
	- REMOVED: Double & Halve checkboxes. 
		- You can still use the Double (x2) and Halve (1/2) buttons to change each dimension, or SHIFT-CLICK them to update all dimensions together.
	Behind the Scenes:
	- Minimap Icon is better and more adaptable to if you use other minimaps / UI's
	- Improved ToolTips & UI Indications when things are disabled
	- Improved Slider Logic
	- Keybindings are sorted into their own category to be easier to find

----------------------------------------

v6.0.0 (March 21, 2020)
	- Auto Update Rot. finished and activated  (note: ".go pitch/roll/turn" will be off until the next server mini-update due to syntax issues in the old reply - it will automatically work perfect once the server updates to the new command syntax!)
	- Major restructure of the ChatFilter to clean it up. It should be way more reliable. No more "My commands aren't showing in chat anymore wtf?" hopefully.
	- Move Relative now uses the server commands for best performance
	- More Cowbell

----------------------------------------

v5.0.0 (September 21, 2017)
	- New: Move Relative option - When on, moves the object relative to your current orientation, not the objects. (I.e., object is facing east, you're facing north - 'forward' will move the object north, not east). Credits to shadowbunny88 for the initial idea and code.
	- New: Save / Load Preset Systems for Object Parameters and Rotations.
	- New: Apply rotation button! No longer need to move sliders around to apply rotation to a new object.
	- New: Hover your Mouse over almost any Text EditBox / CheckBox / Button for more details on it's function or usage.
	- New: Half / Double buttons per-dimension & scale. These are not dynamic and just update the box completely. Works alongside dynamic Double and Halve options still.
	- Added: Ability to set Key Bindings for toggling Move Object and Move Relative Options
	- Added: Lock Frame button besides the close button - When enabled, the frame cannot be moved.
	- Changed: Default 'Step' for Sliders is now in Whole Numbers (Use SHIFT or ALT modifiers to get decimal points), text-entry will auto-round to 4 decimal places.
	- Changed / Fixed: Scale will now wait 0.5 seconds before setting to avoid a server bug with scales set at same time as spawn not saving.
	- Fixed: Bindings for Halve / Bifold toggle.
	- Fixed: Options set via Binding not saving.
	- Fixed: Scale CheckButton un-checking if you are entering a new scale (now only disables the box when an invalid scale is present, re-enables when valid)
	- Not-Fixed: Compatibility with WIM (chat filter counting) (WIM Breaks everything. Don't use it.)
	- WIP: Auto Update Rot. - Button is added but non-functional as there's no feasible way to get the information needed from the server currently.

	
----------------------------------------

v4.2.0 (Aug 17, 2017)
	- Fixed Up / Down not working until Forward / Back was used once
	- Fixed Up / Down not using the correct number (again, sorry - ACTUALLY FIXED NOW!)
	- Changed saved variables into a table instead of a different saved variable for each one - allows new variables to be added on the fly, and provided some work on using tables and how they work so I can try and make auto-dimensions work later.
	- (b) Fixed a removed function being called, resulting in a Lua Error.

----------------------------------------

v4.1.1 (Aug 7, 2017)
	- Fixed Left / Right & Up / Down using their correct numbers instead of just always using the Length Value. Oops..
	- Introduced Temporary Rotation Step Increment Modifiers: Hold Shift to snap to full degrees, alt to snap to 0.1 degrees, both Alt+Shift to snap to every 5 degrees. Default without any modifiers is still 0.01 degrees.

----------------------------------------

v4.0.0 (Aug 5, 2017)
	- Remade all movement functions to utilize ".gps <dir> <unit>".
	- Remade Chat Clarifiers to filter and gather data from chat messages.
	- Removed old deprecated functions and options (i.e., Save on Spawn)
	- Remade entire double, half, and scale modifiers for movement.
	- Expanded to include "Move Object" instead
	- Expanded to include new GObject Rotation sliders.
	- Saved Variables to recall what options and settings you had from last login.
	- Redid UI Color / Transparency / Lightness
	- Made UI Moveable.
	- Made Minimap Button Save position
	- Attempted to start Auto Dimensions System, TBD Later if it'll come to fruition.
	- Changed Keybindings to utilize data validation on object parameters.
	
----------------------------------------

v1.0-3.99
	- All Credits to Dkcaptain for the original addon.
	
----------------------------------------

ObjectMover/HTMLUtils.lua
local directReplacements = {
	["/col"] = "|r",
};

local function convertTextTag(tag)

	if directReplacements[tag] then -- Direct replacement
		return directReplacements[tag];
	elseif tag:match("^col%:%a$") then -- Color replacement
		return Utils.str.color(tag:match("^col%:(%a)$"));
	elseif tag:match("^col:%x%x%x%x%x%x$") then -- Hexa color replacement
		return "|cff"..tag:match("^col:(%x%x%x%x%x%x)$");
	elseif tag:match("^icon%:[^:]+%:%d+$") then -- Icon
		local icon, size = tag:match("^icon%:([^:]+)%:(%d+)$");
		return Utils.str.icon(icon, size);
	end

	return "{"..tag.."}";
end

local function convertTextTags(text)
	if text then
		text = text:gsub("%{(.-)%}", convertTextTag);
		return text;
	end
end

local escapedHTMLCharacters = {
	["<"] = "&lt;",
	[">"] = "&gt;",
	["\""] = "&quot;",
};

local structureTags = {
	["{h(%d)}"] = "<h%1>",
	["{h(%d):c}"] = "<h%1 align=\"center\">",
	["{h(%d):r}"] = "<h%1 align=\"right\">",
	["{/h(%d)}"] = "</h%1>",

	["{p}"] = "<P>",
	["{p:c}"] = "<P align=\"center\">",
	["{p:r}"] = "<P align=\"right\">",
	["{/p}"] = "</P>",
};

--- alignmentAttributes is a conversion table for taking a single-character
--  alignment specifier and getting a value suitable for use in the HTML
--  "align" attribute.
local alignmentAttributes = {
	["c"] = "center",
	["l"] = "left",
	["r"] = "right",
};

--- IMAGE_PATTERN is the string pattern used for performing image replacements
--  in strings that should be rendered as HTML.
---
--- The accepted form this is "{img:<src>:<width>:<height>[:align]}".
---
--- Each individual segment matches up to the next present colon. The third
--- match (height) and everything thereafter needs to check up-to the next
--- colon -or- ending bracket since they could be the final segment.
---
--- Optional segments should of course have the "?" modifer attached to
--- their preceeding colon, and should use * for the content match rather
--- than +.
local IMAGE_PATTERN = [[{img%:([^:]+)%:([^:]+)%:([^:}]+)%:?([^:}]*)%}]];

--- Note that the image tag has to be outside a <P> tag.
---@language HTML
local IMAGE_TAG = [[</P><img src="%s" width="%s" height="%s" align="%s"/><P>]];

-- Convert the given text by his HTML representation


stringtoHTML = function(text, noColor)

	local linkColor = "|cff00ff00";
	if noColor then
		linkColor = "";
	end

	-- 1) Replacement : & character
	text = text:gsub("&", "&amp;");

	-- 2) Replacement : escape HTML characters
	for pattern, replacement in pairs(escapedHTMLCharacters) do
		text = text:gsub(pattern, replacement);
	end

	-- 3) Replace Markdown
	local titleFunction = function(titleChars, title)
		local titleLevel = #titleChars;
		return "\n<h" .. titleLevel .. ">" .. strtrim(title) .. "</h" .. titleLevel .. ">";
	end;

	text = text:gsub("^(#+)(.-)\n", titleFunction);
	text = text:gsub("\n(#+)(.-)\n", titleFunction);
	text = text:gsub("\n(#+)(.-)$", titleFunction);
	text = text:gsub("^(#+)(.-)$", titleFunction);

	-- 4) Replacement : text tags
	for pattern, replacement in pairs(structureTags) do
		text = text:gsub(pattern, replacement);
	end

	local tab = {};
	local i=1;
	while text:find("<") and i<500 do

		local before;
		before = text:sub(1, text:find("<") - 1);
		if #before > 0 then
			tinsert(tab, before);
		end

		local tagText;

		local tag = text:match("</(.-)>");
		if tag then
			tagText = text:sub( text:find("<"), text:find("</") + #tag + 2);
			if #tagText == #tag + 3 then
				return loc.PATTERN_ERROR;
			end
			tinsert(tab, tagText);
		else
			return loc.PATTERN_ERROR;
		end

		local after;
		after = text:sub(#before + #tagText + 1);
		text = after;

		--- 	Log.log("Iteration "..i);
		--- 	Log.log("before ("..(#before).."): "..before);
		--- 	Log.log("tagText ("..(#tagText).."): "..tagText);
		--- 	Log.log("after ("..(#before).."): "..after);

		i = i+1;
		if i == 500 then
			log("HTML overfloooow !", Log.level.SEVERE);
		end
	end
	if #text > 0 then
		tinsert(tab, text); -- Rest of the text
	end

	--- log("Parts count "..(#tab));

	local finalText = "";
	for _, line in pairs(tab) do

		if not line:find("<") then
			line = "<P>" .. line .. "</P>";
		end
		line = line:gsub("\n","<br/>");

		-- Image tag. Specifiers after the height are optional, so they
		-- must be suitably defaulted and validated.
		line = line:gsub(IMAGE_PATTERN, function(img, width, height, align)
			-- If you've not given an alignment, or it's entirely invalid,
			-- you'll get the old default of center.
			align = alignmentAttributes[align] or "center";

			-- Don't blow up on non-numeric inputs. They won't display properly
			-- but that's a separate issue.
			width = tonumber(width) or 128;
			height = tonumber(height) or 128;

			-- Width and height should be absolute.
			-- The tag accepts negative value but people used that to fuck up their profiles
			return string.format(IMAGE_TAG, img, math.abs(width), math.abs(height), align);
		end);

		line = line:gsub("%!%[(.-)%]%((.-)%)", function(icon, size)
			if icon:find("\\") then
				-- If icon text contains \ we have a full texture path
				local width, height;
				if size:find("%,") then
					width, height = strsplit(",", size);
				else
					width = tonumber(size) or 128;
					height = width;
				end
				-- Width and height should be absolute.
				-- The tag accepts negative value but people used that to fuck up their profiles
				return string.format(IMAGE_TAG, icon, math.abs(width), math.abs(height), "center");
			end
			return Utils.str.icon(icon, tonumber(size) or 25);
		end);

		line = line:gsub("%[(.-)%]%((.-)%)",
			"<a href=\"%2\">" .. linkColor .. "[%1]|r</a>");

		line = line:gsub("{link%*(.-)%*(.-)}",
			"<a href=\"%1\">" .. linkColor .. "[%2]|r</a>");

		line = line:gsub("{twitter%*(.-)%*(.-)}",
			"<a href=\"twitter%1\">|cff61AAEE%2|r</a>");

		finalText = finalText .. line;
	end

	finalText = convertTextTags(finalText);

	return "<HTML><BODY>" .. finalText .. "</BODY></HTML>";
end

ObjectMover/Changelog.lua
ObjectMoverChangelog = [[
# v7.0.1 (August 3, 2021)

	- Added an in-game Changelog/Help Manual (Options tab added but not functional yet)
	- Made the main window respect ALT+Z + Better UI Frame
	- Made the main window resizable (Grabby button in the bottom right - |cffFFAAAARight-Click|r to reset the size to default)
	
	- Reminded everyone that "Double" and "Half" features are still here. You can |cffFFAAAAShift-Click|r on the x2 and 1/2 buttons on Length/Width/Height in order to update all three at the same time, which is exactly what the old checkboxes did in the background.

## __________________________________________________

# v7.0.0 (July 29, 2021)

	- NEW: Added Tint & Object Spell Support
	- NEW: Object Group Support (Automatic)
	- NEW: Automatic Dimension Detection for M2's
	- NEW: Auto Update mode for Object Parameters (Will auto fill selected objects ID, Scale, and if it's an M2, the dimensions as well)
	- NEW: Auto Show option - Toggle on to automatically show ObjectMover when you load in
	- NEW: Fade Out option - Toggle on to fade ObjectMover to be less distracting when not in immediate use
	- UPDATED: Scale Object now works with Macro Spawning instead of delayed scale commands (No Delay, No Desync)
	- UPDATED: Scale Object now applies to the selected object when entered
	- UPDATED: Saved Presets can now be deleted directly from the UI (Right-Click in Load Dropdown to delete, no need to use chat commands now)
	- CHANGED: Spawn is now in Object section; Revised Movement button placements
	- CHANGED: Added a "Classic Layout" button in Movement to revert the movement buttons to closer to the original layout
	- REMOVED: Double & Halve checkboxes. 
		- You can still use the Double (x2) and Halve (1/2) buttons to change each dimension, or SHIFT-CLICK them to update all dimensions together.
	Behind the Scenes:
	- Minimap Icon is better and more adaptable to if you use other minimaps / UI's
	- Improved ToolTips & UI Indications when things are disabled
	- Improved Slider Logic
	- Keybindings are sorted into their own category to be easier to find

## __________________________________________________

# v6.0.0 (March 21, 2020)

	- Auto Update Rot. finished and activated  (note: ".go pitch/roll/turn" will be off until the next server mini-update due to syntax issues in the old reply - it will automatically work perfect once the server updates to the new command syntax!)
	- Major restructure of the ChatFilter to clean it up. It should be way more reliable. No more "My commands aren't showing in chat anymore wtf?" hopefully.
	- Move Relative now uses the server commands for best performance
	- More Cowbell
	
## __________________________________________________

See [https://forums.epsilonwow.net/topic/467-addon-objectmover/](https://forums.epsilonwow.net/topic/467-addon-objectmover/) for more.
]]

ObjectMoverHelpManual = [[
# Object Mover Manual

## Object Info

### Parameters:
|cffFFCC00 - Object ID|r - This is where you would fill in the ID of the Object you want to spawn. Only single objects are supported here, not Blueprints.
|cffFFCC00 - Length/Width/Height|r - This determines how far you, or the object, will move when using the Movement buttons. AKA: The object dimensions.
|cffFFCC00 - Scale|r - Changing this will scale your currently selected object, and, |cffFFAAAAif the checkbox is enabled|r, apply the scale to any newly spawned objected. If the checkbox is enabled, all Length/Width/Height will be automatically adjusted for the scale as well. |cffAAAAAA(Example: Lenght 4 at Scale 2 will move by 8.)|r

### Buttons:
|cffFFCC00 - Get ID|r - This will get the ID of the currently selected Object and fill it into the Object ID box. |cffFFAAAARight-Click|r will also try and auto-fill the Length/Width/Height parameters.
|cffFFCC00 - x2 (Double)|r - This will double the number in the dimension that it is in line with. |cffFFAAAAShift-Click|r to double all 3 dimensions.
|cffFFCC00 - 1/2 (Half)|r - This will halve the number in the dimension that it is in line with. |cffFFAAAAShift-Click|r to halve all 3 dimensions.
|cffFFCC00 - Save|r - Saves all the current information in the Object Info section to a new pre-set for later use. Leaving something blank, or as 0, will not 'save' that specific setting, allowing you to save just the data you want if needed.
|cffFFCC00 - Load|r - Loads a previously saved pre-set. |cffFFAAAARight-Click|r a pre-set to delete it.
|cffFFCC00 - Auto Update|r - Updates Object ID, Length/Width/Height, and Scale info when selecting an object. Note that Length/Width/Height will only update when selecting an |cffFFAAAAM2 object|r, not a WMO.
|cffFFCC00 - Spawn|r - Spawns the currently selected object. Automatically scales it if Scale is set & checked.

## __________________________________________________

## Movement

|cffFFCC00 - Forward/Back|r - Moves you, or the object, Forward/Back by the 'Length' amount.
|cffFFCC00 - Left/Right|r - Moves you, or the object, Left/Right by the 'Width' amount.
|cffFFCC00 - Up/Down|r - Moves you, or the object, Up/Down by the 'Height' amount.
|cffFFCC00 - Go To|r - Teleports you to the selected object.
|cffFFCC00 - Classic Layout|r - Adjusts the Movement buttons to a more classic (WASD) style layout.

## __________________________________________________

## Options

|cffFFCC00 - Move Object|r - Move the selected Game Object instead of yourself.
|cffFFCC00 - Spawn on Move|r - Spawns an object each time you move - not recommended in |cffFFAAAAMove Object|r mode.
|cffFFCC00 - Move Relative|r - Moves the object relative to the direction YOU are facing instead of the direction the object is facing. Only works in |cffFFAAAAMove Object|r mode. 
|cffFFCC00 - Auto Show|r - Always Show the ObjectMover UI when first loading in.
|cffFFCC00 - Fade Out|r - Fade the UI out when not currently moused over.
|cffFFCC00 - Messages|r - Show Rotation & Tint messages in chat. This will spam your chat.

## __________________________________________________

## Rotation

### Sliders

These will automatically apply to the object as you drag the sliders:
|cffFFCC00 - Roll|r - Rotate/Tilt the object on a side-to-side horizonatal axis.
|cffFFCC00 - Pitch|r - Pitch/Tilt the object on a forward-to-back horizontal axis.
|cffFFCC00 - Turn|r - Rotate the object around it's center on a vertical axis.

|cffFFAAAA - Hold Shift or Alt|r while dragging a slider to adjust it's decimal places to get a more percise rotation.

### Buttons

|cffFFCC00 - Save|r - Save the current rotations as a pre-set. Set a rotation to -1 to ignore it when saving/loading.
|cffFFCC00 - Load|r - Load a saved pre-set. Values saved as -1 will be ignored. |cffFFAAAARight-Click|r a pre-set to delete it.
|cffFFCC00 - Auto Update|r - Updates the Roll/Pitch/Turn sliders to match an object when selected.
|cffFFCC00 - Apply|r - Apply the current rotation to the selected object.

## __________________________________________________

## Tint

|cffFFCC00 - Red/Green/Blue|r - Adjust the tint colors of an object.
|cffFFCC00 - Transparency|r - Adjust the transparency of an object.
|cffFFCC00 - Color Picker|r - Open a color wheel to select a color & transparency from there.
|cffFFCC00 - Apply|r - Apply the current color to the selected object. |cffFFAAAARight-Click|r to remove/reset the tint on an object.
|cffFFCC00 - Spell|r - Apply a spell as a visual effect on an object. |cffFFAAAARight-Click|r to remove/reset the spells on an object.
|cffFFCC00 - Auto Update|r - Updates the tint sliders & spell button to match an object when selected.

]]